File: clients/ios/Lextures/Features/ContentTools/Sandbox/SandboxWebViewHost.swift
import SwiftUI
import WebKit

/// CT.M4 sandboxed tool host: native document fetch → opaque-origin WKWebView → bridge.
struct SandboxWebViewHost: View {
    @Environment(AuthSession.self) private var session
    @Environment(\.locale) private var locale
    @Environment(\.layoutDirection) private var layoutDirection
    @Environment(\.openURL) private var openURL

    let toolId: String
    let instanceId: String
    let toolVersion: String
    let title: String
    let config: JSONValue
    let state: JSONValue
    let revision: Int64
    let readOnly: Bool
    let capabilities: [String]
    var save: (JSONValue) -> Void
    var runAction: (String, JSONValue) async throws -> JSONValue?
    var announce: (String, Bool) -> Void

    @State private var height: CGFloat = 160
    @State private var ready = false
    @State private var timedOut = false
    @State private var crashed = false
    @State private var documentHTML: String?
    @State private var fetchFailed = false
    @State private var reloadToken = 0

    var body: some View {
        Group {
            if crashed {
                ToolErrorCardView(
                    toolName: title,
                    message: L.text("mobile.contentTools.sandbox.crashed"),
                    onRetry: { resetAndReload() }
                )
            } else if timedOut && !ready {
                ToolErrorCardView(
                    toolName: title,
                    message: L.text("mobile.contentTools.sandbox.timeout"),
                    onRetry: { resetAndReload() }
                )
            } else if fetchFailed {
                ToolPlaceholderView(
                    reason: .unavailable,
                    message: L.text("mobile.contentTools.sandbox.needsConnection")
                )
            } else {
                VStack(alignment: .leading, spacing: 4) {
                    if !ready {
                        ToolPlaceholderView(reason: .loading)
                    }
                    if let documentHTML {
                        SandboxWKView(
                            html: documentHTML,
                            instanceId: instanceId,
                            toolId: toolId,
                            config: config,
                            state: state,
                            revision: revision,
                            readOnly: readOnly,
                            locale: locale.identifier,
                            dir: layoutDirection == .rightToLeft ? "rtl" : "ltr",
                            height: $height,
                            ready: $ready,
                            crashed: $crashed,
                            capabilities: capabilities,
                            save: save,
                            runAction: runAction,
                            announce: announce,
                            onOpenURL: { openURL($0) }
                        )
                        .frame(height: ready ? height : 0)
                        .clipped()
                        .opacity(ready ? 1 : 0)
                        .accessibilityElement(children: .contain)
                        .accessibilityLabel("\(title), \(L.text("mobile.contentTools.sandbox.badge"))")
                    }
                    Text(L.text("mobile.contentTools.sandbox.badge"))
                        .font(.caption2)
                        .foregroundStyle(.secondary)
                }
            }
        }
        .task(id: "\(toolId)|\(instanceId)|\(reloadToken)") {
            await fetchDocument()
        }
        .onAppear {
            _ = SandboxWebViewPool.shared.retain(instanceId)
        }
        .onDisappear {
            SandboxWebViewPool.shared.release(instanceId)
        }
    }

    private func resetAndReload() {
        crashed = false
        timedOut = false
        ready = false
        fetchFailed = false
        documentHTML = nil
        reloadToken += 1
    }

    private func fetchDocument() async {
        documentHTML = nil
        fetchFailed = false
        ready = false
        timedOut = false
        let path = ContentToolSandboxLogic.documentPath(toolId: toolId, version: toolVersion)
        let url = AppConfiguration.webURL(path: path)
        var request = URLRequest(url: url)
        if let token = session.accessToken, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, (200 ... 299).contains(http.statusCode),
                  let html = String(data: data, encoding: .utf8), !html.isEmpty else {
                fetchFailed = true
                return
            }
            documentHTML = html
            try? await Task.sleep(nanoseconds: UInt64(ContentToolSandboxLogic.readyTimeoutMs) * 1_000_000)
            if !Task.isCancelled && !ready {
                timedOut = true
            }
        } catch {
            fetchFailed = true
        }
    }
}

// MARK: - WKWebView representable

private struct SandboxWKView: UIViewRepresentable {
    let html: String
    let instanceId: String
    let toolId: String
    let config: JSONValue
    let state: JSONValue
    let revision: Int64
    let readOnly: Bool
    let locale: String
    let dir: String
    @Binding var height: CGFloat
    @Binding var ready: Bool
    @Binding var crashed: Bool
    let capabilities: [String]
    var save: (JSONValue) -> Void
    var runAction: (String, JSONValue) async throws -> JSONValue?
    var announce: (String, Bool) -> Void
    var onOpenURL: (URL) -> Void

    func makeCoordinator() -> Coordinator {
        Coordinator(parent: self)
    }

    func makeUIView(context: Context) -> WKWebView {
        let wkConfig = WKWebViewConfiguration()
        // Non-persistent store — nothing the tool writes survives the mount (FR-2 / privacy).
        wkConfig.websiteDataStore = .nonPersistent()
        wkConfig.preferences.javaScriptCanOpenWindowsAutomatically = false
        wkConfig.defaultWebpagePreferences.allowsContentJavaScript = true

        let webView = WKWebView(frame: .zero, configuration: wkConfig)
        webView.navigationDelegate = context.coordinator
        webView.uiDelegate = context.coordinator
        webView.scrollView.isScrollEnabled = false
        webView.isOpaque = false
        webView.backgroundColor = .clear

        let bridge = SandboxBridge(webView: webView, handlers: context.coordinator.makeHandlers())
        bridge.attach(to: webView.configuration.userContentController)
        context.coordinator.bridge = bridge
        context.coordinator.webView = webView

        // Opaque origin: loadHTMLString with nil baseURL (FR-2).
        webView.loadHTMLString(html, baseURL: nil)
        return webView
    }

    func updateUIView(_ webView: WKWebView, context: Context) {
        context.coordinator.parent = self
    }

    static func dismantleUIView(_ uiView: WKWebView, coordinator: Coordinator) {
        coordinator.bridge?.dispose(controller: uiView.configuration.userContentController)
        coordinator.bridge = nil
        uiView.stopLoading()
        uiView.navigationDelegate = nil
        uiView.uiDelegate = nil
    }

    @MainActor
    final class Coordinator: NSObject, WKNavigationDelegate, WKUIDelegate {
        var parent: SandboxWKView
        var bridge: SandboxBridge?
        weak var webView: WKWebView?
        private var didInit = false

        init(parent: SandboxWKView) {
            self.parent = parent
        }

        func makeHandlers() -> SandboxBridge.Handlers {
            SandboxBridge.Handlers(
                onReady: { [weak self] contract in
                    guard let self else { return }
                    if ContentToolSandboxLogic.contractInSupportedRange(contract) {
                        self.parent.ready = true
                    } else {
                        self.parent.crashed = true
                    }
                },
                onSave: { [weak self] state, _ in
                    guard let self, !self.parent.readOnly else { return }
                    self.parent.save(state)
                    self.bridge?.postStateAccepted(revision: self.parent.revision + 1)
                },
                onRunAction: { [weak self] id, action, input in
                    guard let self else { return }
                    Task { @MainActor in
                        do {
                            let result = try await self.parent.runAction(action, input)
                            self.bridge?.postActionResult(id: id, result: result)
                        } catch {
                            self.bridge?.postError(
                                id: id,
                                code: "action_failed",
                                message: error.localizedDescription
                            )
                        }
                    }
                },
                onResize: { [weak self] height in
                    self?.parent.height = CGFloat(height)
                },
                onAnnounce: { [weak self] message, assertive in
                    self?.parent.announce(message, assertive)
                },
                onInvalid: { _ in },
                onMetric: { _, _ in }
            )
        }

        func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
            guard !didInit else { return }
            didInit = true
            bridge?.postInit(
                instanceId: parent.instanceId,
                config: parent.config,
                state: parent.state,
                revision: parent.revision,
                locale: parent.locale,
                dir: parent.dir,
                readOnly: parent.readOnly,
                participantId: ContentToolSandboxLogic.opaqueParticipantId(parent.instanceId)
            )
        }

        func webViewWebContentProcessDidTerminate(_ webView: WKWebView) {
            parent.crashed = true
        }

        func webView(
            _ webView: WKWebView,
            decidePolicyFor navigationAction: WKNavigationAction,
            decisionHandler: @escaping (WKNavigationActionPolicy) -> Void
        ) {
            // Allow the initial about:blank / html load; block everything else (FR-13).
            if navigationAction.navigationType == .other {
                let url = navigationAction.request.url
                if url == nil || url?.absoluteString == "about:blank" || url?.scheme == nil {
                    decisionHandler(.allow)
                    return
                }
            }
            if let url = navigationAction.request.url {
                parent.onOpenURL(url)
            }
            decisionHandler(.cancel)
        }

        func webView(
            _ webView: WKWebView,
            createWebViewWith configuration: WKWebViewConfiguration,
            for navigationAction: WKNavigationAction,
            windowFeatures: WKWindowFeatures
        ) -> WKWebView? {
            if let url = navigationAction.request.url {
                parent.onOpenURL(url)
            }
            return nil
        }

        func webView(
            _ webView: WKWebView,
            requestMediaCapturePermissionFor origin: WKSecurityOrigin,
            initiatedByFrame frame: WKFrameInfo,
            type: WKMediaCaptureType,
            decisionHandler: @escaping (WKPermissionDecision) -> Void
        ) {
            let needed: String
            switch type {
            case .camera: needed = "camera"
            case .microphone: needed = "microphone"
            case .cameraAndMicrophone: needed = "camera"
            @unknown default:
                decisionHandler(.deny)
                return
            }
            decisionHandler(parent.capabilities.contains(needed) ? .prompt : .deny)
        }
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/features/contenttools/sandbox/SandboxWebViewHost.kt
package com.lextures.android.features.contenttools.sandbox

import android.annotation.SuppressLint
import android.net.Uri
import android.view.ViewGroup
import android.webkit.PermissionRequest
import android.webkit.WebChromeClient
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.viewinterop.AndroidView
import com.lextures.android.R
import com.lextures.android.core.design.textSecondary
import com.lextures.android.core.i18n.L
import com.lextures.android.core.lms.ContentToolSandboxLogic
import com.lextures.android.core.network.ApiClient
import com.lextures.android.features.contenttools.ToolErrorCard
import com.lextures.android.features.contenttools.ToolPlaceholder
import com.lextures.android.features.contenttools.ToolPlaceholderReason
import java.util.Locale
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

/**
 * CT.M4 sandboxed tool host: native document fetch → opaque-origin WebView → bridge.
 */
@SuppressLint("SetJavaScriptEnabled")
@Composable
fun SandboxWebViewHost(
    toolId: String,
    instanceId: String,
    toolVersion: String,
    title: String,
    config: JsonElement,
    state: JsonElement,
    revision: Long,
    readOnly: Boolean,
    capabilities: List<String>,
    accessToken: String?,
    save: (JsonObject) -> Unit,
    runAction: suspend (name: String, input: JsonObject) -> JsonElement?,
    announce: (String, Boolean) -> Unit,
    onOpenUrl: (Uri) -> Unit = {},
    modifier: Modifier = Modifier,
) {
    val scope = rememberCoroutineScope()
    val configuration = LocalConfiguration.current
    val localeTag = remember(configuration) {
        configuration.locales[0]?.toLanguageTag() ?: Locale.getDefault().toLanguageTag()
    }
    val dir = if (configuration.layoutDirection == android.util.LayoutDirection.RTL) "rtl" else "ltr"

    var heightDp by remember { mutableStateOf(160.0) }
    var ready by remember(instanceId) { mutableStateOf(false) }
    var timedOut by remember(instanceId) { mutableStateOf(false) }
    var crashed by remember(instanceId) { mutableStateOf(false) }
    var documentHtml by remember(instanceId) { mutableStateOf<String?>(null) }
    var fetchFailed by remember(instanceId) { mutableStateOf(false) }
    var reloadToken by remember(instanceId) { mutableStateOf(0) }
    var bridgeRef by remember { mutableStateOf<SandboxBridge?>(null) }
    var unsupportedBridge by remember { mutableStateOf(false) }

    val badge = L.text(R.string.mobile_contentTools_sandbox_badge)
    val timeoutMsg = L.text(R.string.mobile_contentTools_sandbox_timeout)
    val crashedMsg = L.text(R.string.mobile_contentTools_sandbox_crashed)
    val needsConnection = L.text(R.string.mobile_contentTools_sandbox_needsConnection)
    val a11y = "$title, $badge"

    fun resetAndReload() {
        crashed = false
        timedOut = false
        ready = false
        fetchFailed = false
        documentHtml = null
        unsupportedBridge = false
        bridgeRef?.dispose()
        bridgeRef = null
        reloadToken += 1
    }

    DisposableEffect(instanceId) {
        SandboxWebViewPool.retain(instanceId)
        onDispose {
            bridgeRef?.dispose()
            bridgeRef = null
            SandboxWebViewPool.release(instanceId)
        }
    }

    LaunchedEffect(toolId, instanceId, reloadToken, accessToken) {
        documentHtml = null
        fetchFailed = false
        ready = false
        timedOut = false
        val path = ContentToolSandboxLogic.documentPath(toolId, toolVersion)
        val html = runCatching {
            val (body, code) = ApiClient().request(
                path = path,
                method = "GET",
                accessToken = accessToken,
            )
            if (code in 200..299) body.takeIf { it.isNotBlank() } else null
        }.getOrNull()
        if (html == null) {
            fetchFailed = true
            return@LaunchedEffect
        }
        documentHtml = html
        delay(ContentToolSandboxLogic.READY_TIMEOUT_MS.toLong())
        if (!ready) timedOut = true
    }

    when {
        unsupportedBridge -> {
            ToolPlaceholder(
                reason = ToolPlaceholderReason.OPEN_IN_BROWSER,
                toolName = title,
                message = needsConnection,
                modifier = modifier,
            )
        }
        crashed -> {
            ToolErrorCard(
                toolName = title,
                message = crashedMsg,
                onRetry = { resetAndReload() },
                modifier = modifier,
            )
        }
        timedOut && !ready -> {
            ToolErrorCard(
                toolName = title,
                message = timeoutMsg,
                onRetry = { resetAndReload() },
                modifier = modifier,
            )
        }
        fetchFailed -> {
            ToolPlaceholder(
                reason = ToolPlaceholderReason.UNAVAILABLE,
                message = needsConnection,
                modifier = modifier,
            )
        }
        else -> {
            Column(modifier.fillMaxWidth()) {
                if (!ready) {
                    ToolPlaceholder(reason = ToolPlaceholderReason.LOADING)
                }
                val html = documentHtml
                if (html != null) {
                    AndroidView(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(if (ready) heightDp.dp else 0.dp)
                            .semantics { contentDescription = a11y },
                        factory = { ctx ->
                            WebView(ctx).apply {
                                layoutParams = ViewGroup.LayoutParams(
                                    ViewGroup.LayoutParams.MATCH_PARENT,
                                    ViewGroup.LayoutParams.WRAP_CONTENT,
                                )
                                settings.javaScriptEnabled = true
                                settings.domStorageEnabled = false
                                settings.allowFileAccess = false
                                settings.allowContentAccess = false
                                // Opaque origin (FR-2): loadDataWithBaseURL(null, …)
                                webViewClient = object : WebViewClient() {
                                    override fun shouldOverrideUrlLoading(
                                        view: WebView?,
                                        request: WebResourceRequest?,
                                    ): Boolean {
                                        val uri = request?.url
                                        if (uri != null) onOpenUrl(uri)
                                        return true // block all navigation (FR-13)
                                    }

                                    override fun onPageFinished(view: WebView?, url: String?) {
                                        val bridge = SandboxBridge.create(
                                            this@apply,
                                            SandboxBridge.Handlers(
                                                onReady = { contract ->
                                                    if (ContentToolSandboxLogic.contractInSupportedRange(contract)) {
                                                        ready = true
                                                    } else {
                                                        crashed = true
                                                    }
                                                },
                                                onSave = { nextState, _ ->
                                                    if (!readOnly) {
                                                        val obj = nextState as? JsonObject
                                                            ?: JsonObject(emptyMap())
                                                        save(obj)
                                                        bridgeRef?.postStateAccepted(revision + 1)
                                                    }
                                                },
                                                onRunAction = { id, action, input ->
                                                    scope.launch {
                                                        try {
                                                            val result = runAction(
                                                                action,
                                                                input as? JsonObject ?: JsonObject(emptyMap()),
                                                            )
                                                            bridgeRef?.postActionResult(id, result)
                                                        } catch (e: Exception) {
                                                            bridgeRef?.postError(
                                                                id,
                                                                "action_failed",
                                                                e.message ?: "action failed",
                                                            )
                                                        }
                                                    }
                                                },
                                                onResize = { h -> heightDp = h },
                                                onAnnounce = announce,
                                            ),
                                        )
                                        if (bridge == null) {
                                            unsupportedBridge = true
                                            return
                                        }
                                        bridgeRef = bridge
                                        bridge.postInit(
                                            instanceId = instanceId,
                                            config = config,
                                            state = state,
                                            revision = revision,
                                            locale = localeTag,
                                            dir = dir,
                                            readOnly = readOnly,
                                            participantId = ContentToolSandboxLogic.opaqueParticipantId(instanceId),
                                        )
                                    }
                                }
                                webChromeClient = object : WebChromeClient() {
                                    override fun onPermissionRequest(request: PermissionRequest?) {
                                        val req = request ?: return
                                        val granted = req.resources.filter { resource ->
                                            when (resource) {
                                                PermissionRequest.RESOURCE_VIDEO_CAPTURE ->
                                                    "camera" in capabilities
                                                PermissionRequest.RESOURCE_AUDIO_CAPTURE ->
                                                    "microphone" in capabilities
                                                else -> false
                                            }
                                        }.toTypedArray()
                                        if (granted.isEmpty()) req.deny() else req.grant(granted)
                                    }

                                    override fun onCreateWindow(
                                        view: WebView?,
                                        isDialog: Boolean,
                                        isUserGesture: Boolean,
                                        resultMsg: android.os.Message?,
                                    ): Boolean = false
                                }
                                loadDataWithBaseURL(null, html, "text/html", "UTF-8", null)
                            }
                        },
                        update = { /* bridge owns messaging */ },
                    )
                }
                Text(text = badge, fontSize = 11.sp, color = textSecondary())
            }
        }
    }
}
